packages/@aws-cdk/aws-lambda/lib/function-base.ts

```diff
@@ -431,6 +435,35 @@ export abstract class FunctionBase extends Resource implements IFunction, ec2.IC
     throw new Error(`Invalid principal type for Lambda permission statement: ${principal.constructor.name}. ` +
       'Supported: AccountPrincipal, ArnPrincipal, ServicePrincipal');
   }
+
+  private parseConditions(principal: iam.IPrincipal): { sourceAccount: string, sourceArn: string } | null {
+    if ('conditions' in principal) {
+      const conditions: iam.Conditions = (principal as iam.PrincipalWithConditions).policyFragment.conditions;
+      const conditionPairs = Object.entries(conditions)
+        .reduce(
```

Comment:
Personally I'm more a fan of writing this in the form of `flatMap(xs, x => [...])` or `flatten(xs.map(x => [...]))`.

`reduces` always seem unreadable to me.

`flatten` is easily implemented as:

```ts
function flatten<A>(xs: A[][]): A[] {
  return Array.prototype.concat.apply([], xs);
}
```

Reply:
I originally went with `reduce` to avoid re-writing `flatten` or `flatMap` for the nth time in the repo. Not sure if there is a better way to avoid code duplication in the poly repo, but maybe this can be cleaned up in v2?

Reply:
We don't have a `utils` package on purpose. Yes, it's a bit more code, but on the other hand packages are nice and self-contained, usually.